[quant] Add tensor_qparam variant to fake_quantize_per_tensor

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61318
* **#61317**

Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

supriyar committed:
[quant] Add tensor_qparam variant to fake_quantize_per_tensor

Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

raghuramank100 commented:
Looks good to me. Will we deprecate the existing fake-quant API in the next release? Seems redundant to have both.

raghuramank100 reviewed (approved):
Looks good!. Wondering if we should plan for deprecating the current fake-quant API.  Would it be possible to do that in a later PR?

supriyar commented:
> Looks good!. Wondering if we should plan for deprecating the current fake-quant API. Would it be possible to do that in a later PR?

The older one is a public facing API though, so I don't think we can fully remove it without breaking BC.

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2089,6 +2090,33 @@ void fake_quantize_tensor_cachemask_kernel(
+  float* scale_ptr = sc.data_ptr<float>();
+  int32_t* zp_ptr = z_point.data_ptr<int32_t>();
+
+  _fake_quantize_tensor_helper(output, mask, input, *scale_ptr, *zp_ptr, quant_min, quant_max);
```
is this trying to get the per tensor scale parameter? I think we can use sc.item() since it's a scalar

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5223,11 +5223,20 @@
   dispatch:
     CPU, CUDA: fake_quantize_per_tensor_affine_cachemask
 
+- func: _fake_quantize_per_tensor_affine_cachemask_tensor_qparams(Tensor self, Tensor scale, Tensor zero_point, int quant_min, int quant_max) -> (Tensor output, Tensor mask)
```
why don't we have fake_quantize_per_tensor_affine_cachemask.tesnor_qparams?

supriyar replied:
I kept getting undefined function error. I think it doesn't play well with the automatic selection of the backward function.
Hopefully prefixing it with `_` removes it from the user facing API.

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add tensor_qparam variant to fake_quantize_per_tensor"


Add an overload to fake_quantize_per_tensor that accepts scale/zero_point as input. The reasons to do this are

* required for fused observer + fake_quant operator on GPU where the scale/zero_point will be calculated by the observer on device. Passing tensor inputs enables us to directly access the scale/zero-point value in the cuda kernel to avoid extra copies/malloc
* enables us to pass in float as scale dtype and int32 as zero_point dtype (which is consistent with what the quantize call actually uses) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L52-L53
* overload consistent with `quantizer_per_tensor.tensor_qparams`

Differential Revision: [D29552727](https://our.internmc.facebook.com/intern/diff/D29552727/)

[ghstack-poisoned]
